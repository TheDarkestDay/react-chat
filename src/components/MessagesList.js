import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';

import { Message, User } from '../constants/shapes';
import ChatMessage from './ChatMessage';
import StatusMessage from './StatusMessage';

const chatHeaderHeight = '64px';
const newMessageFormHeight = '64px';

const styles = () => ({
  messagesList: {
    overflow: 'auto',
    maxHeight: `calc(100vh - ${chatHeaderHeight} - ${newMessageFormHeight})`,
    flexGrow: 1,
  },
});

class MessagesList extends Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    messages: PropTypes.arrayOf(Message).isRequired,
    user: User.isRequired,
  }

  componentWillReceiveProps() {
    if (this.messagesList) {
      requestAnimationFrame(() => this.messagesList.scrollTo(0, this.messagesList.scrollHeight));
    }
  }

  initMessagesList = (messagesListRef) => {
    this.messagesList = messagesListRef;
  }

  render() {
    const { classes, messages, user } = this.props;

    return (
      <section className={classes.messagesList} ref={this.initMessagesList}>
        {
          messages.map(message => (
            message.statusMessage
              ? <StatusMessage key={message._id} {...message} />
              : <ChatMessage key={message._id} user={user} {...message} />
          ))
        }
      </section>
    );
  }
}

export default withStyles(styles)(MessagesList);
