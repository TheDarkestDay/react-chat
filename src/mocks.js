/* eslint-disable import/prefer-default-export */
export const userMock = {
  _id: '1',
  username: 'test_user',
};
