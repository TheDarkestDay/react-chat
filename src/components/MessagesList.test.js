/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import MessagesList from './MessagesList';

import { userMock } from '../mocks';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MessagesList messages={[]} user={userMock} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('should match snapshot', () => {
  const tree = renderer.create(<MessagesList messages={[]} user={userMock} />).toJSON();
  expect(tree).toMatchSnapshot();
});
