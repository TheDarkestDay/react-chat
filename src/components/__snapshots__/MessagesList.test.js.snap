// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<section
  className="MessagesList-messagesList-2"
/>
`;
